core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,268 @@
+            .newUpdater(DefaultAsyncLoader.class, RefreshingFuture.class, "loadFuture");
+    private static final ExecutorService DEFAULT_REFRESH_EXECUTOR = Executors.newSingleThreadExecutor();
+
+    private final Function<@Nullable T, CompletableFuture<T>> loader;
+    @Nullable
+    private final Duration expireAfterLoad;
+    @Nullable
+    private final Predicate<@Nullable T> expireIf;
+    @Nullable
+    private final Predicate<@Nullable T> refreshIf;
+    @Nullable
+    private final ExecutorService refreshExecutor;
+    @Nullable
+    private final BiFunction<Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler;
+
+    private volatile RefreshingFuture<T> loadFuture = RefreshingFuture.completedFuture();
+
+    DefaultAsyncLoader(Function<@Nullable T, CompletableFuture<T>> loader,
+                       @Nullable Duration expireAfterLoad,
+                       @Nullable Predicate<@Nullable T> expireIf,
+                       @Nullable Predicate<@Nullable T> refreshIf,
+                       @Nullable ExecutorService refreshExecutor,
+                       @Nullable BiFunction<
+                               Throwable, @Nullable T, @Nullable CompletableFuture<T>> exceptionHandler) {
+        requireNonNull(loader, "loader");
+        this.loader = loader;
+        this.expireAfterLoad = expireAfterLoad;
+        this.expireIf = expireIf;
+        this.refreshIf = refreshIf;
+        this.refreshExecutor = refreshExecutor;
+        this.exceptionHandler = exceptionHandler;
+    }
+
+    @Override
+    public CompletableFuture<T> get() {
+        return get0().thenApply(f -> f.loadVal);
+    }
+
+    private CompletableFuture<CacheEntry<T>> get0() {
+        RefreshingFuture<T> future;
+        CacheEntry<T> cacheEntry = null;
+        boolean needsRefresh = false;
+        for (;;) {
+            final RefreshingFuture<T> loadFuture = this.loadFuture;
+            if (!loadFuture.isDone()) {
+                return loadFuture;
+            }
+
+            if (!loadFuture.isCompletedExceptionally()) {
+                cacheEntry = loadFuture.join();
+                if (isValid(cacheEntry)) {
+                    needsRefresh = needsRefresh(cacheEntry);
+                    if (!needsRefresh || loadFuture.refreshing) {
+                        return loadFuture;
+                    }
+                }
+            }
+
+            future = new RefreshingFuture<>(isValid(cacheEntry) ? cacheEntry : null);
+            if (loadFutureUpdater.compareAndSet(this, loadFuture, future)) {
+                break;
+            }
+        }
+
+        final T cache = cacheEntry != null ? cacheEntry.loadVal : null;
+        if (needsRefresh) {
+            final CompletableFuture<CacheEntry<T>> newRefreshFuture = future;
+            CompletableFuture.runAsync(() -> load(cache, newRefreshFuture),
+                                       refreshExecutor != null ? refreshExecutor : DEFAULT_REFRESH_EXECUTOR);
```

Comment:
Initially, I thought we could use `refreshExecutor` for both the first fetch and refresh. That said, `loader` function returns `CompletableFuture` so users can easily change the working thread with `XXXAsync` methods in `CompletableFuture`.

Should we remove `refreshExecutor` and call `load()` in the current thread? Let's add it later when there is a request for it.
```java
final T cache = cacheEntry != null ? cacheEntry.loadVal : null;
load(cache, future);
```

Reply:
✅ updated, remove `refreshExecutor`